### ISSUE TITLE
Allow to delete Error-state instances

### ISSUE THREAD

**Issue by ddpolyakov** [terraform-providers/terraform-provider-openstack#404]
Hi there, I am running OpenStack Ocata version and deploying infrastructure via Terraform.
When I create instances and they got into the error state for some reasons - I cannot destroy them with terraform. 
I guess there are some issues in handling instance's power state. At Error message, Terraform complains about 
```Invalid current power_state:ERROR ```
but actual power_state of instance in reproduce steps - is RUNNING.

 

### Terraform Version
Terraform v0.11.7
+ provider.openstack v1.8.0

### Affected Resource(s)
- instance resource

### Expected Behavior
delete all error state instances

### Steps to Reproduce

1. create an instance via terraform any way you like, apply it

2. set instance state into "ERROR" using 
```openstack server set --state error  ee21f319-eab1-403a-ae33-5c88bd0a787a```

3.  terraform destroy will fail with an error

```Error: Error refreshing state: 1 error(s) occurred:

* openstack_compute_instance_v2.instance: 1 error(s) occurred:

* openstack_compute_instance_v2.instance[0]: openstack_compute_instance_v2.instance.0: Error changing the instance(ee21f319-eab1-403a-ae33-5c88bd0a787a) power_state. Invalid current power_state:ERROR 




Error: Error refreshing state: 1 error(s) occurred:

* openstack_compute_instance_v2.instance: 1 error(s) occurred:

* openstack_compute_instance_v2.instance[0]: openstack_compute_instance_v2.instance.0: Error changing the instance(ee21f319-eab1-403a-ae33-5c88bd0a787a) power_state. Invalid current power_state:ERROR ```

**Comment by ddpolyakov** [comment]
As far as I discovered - power states are just copying from vm.Status 
https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_compute_instance_v2.go
```	//Set the current power_state:
	if strings.ToLower(server.Status) != "active" && strings.ToLower(server.Status) != "shutoff" {
		return fmt.Errorf("Error changing the instance(%s) power_state. Invalid current power_state:%s ", d.Id(), server.Status)
	} else {
		d.Set("power_state", strings.ToLower(server.Status))
	}
```
and no actual power state handling // mapping is present.

Are there any plans to implement it?

**Comment by jtopjian** [comment]
> Are there any plans to implement it?

Possibly. We need to determine a way to implement this in a safe way that will cover most users' use-cases. For now, this is expected behavior.

`ERROR` states are tricky because there can be a myriad of causes. Almost always, the cause for the error was done out-of-band of Terraform. So to have Terraform support recovery of something that happened out-of-band is a really difficult thing to do.

**Comment by ddpolyakov** [comment]
@jtopjian 
Thanks a lot for a quick responce.
Resolving ```ERROR``` state is too much complicated and for sure it is a cloud provider responsibility in 99% of use-cases. But seemes that error state handling in case of _destroying_ infrastructure - is a bit out-of-scope of resolving errors - it is just deleting resources.

**Comment by jtopjian** [comment]
> But seemes that error state handling in case of destroying infrastructure - is a bit out-of-scope of resolving errors - it is just about deleting resources.

What happens if the instance is not able to be cleanly deleted due to the error state?

**Comment by hardys** [comment]
I ran into this today, and (coming from working on OpenStack, including a lot of work on the Heat orchestration tool) I can say this is pretty common, and in all cases I'm aware of attempting to delete an ERROR state instance isn't likely to cause problems, and in many/most cases it will succeed and allow the destroy cleanup to continue.

Very common reasons for this are running out of resources on the compute nodes & incorrect specification for nodes in the flavor - both can be safely deleted, and if the error is with the service itself then the delete could fail and IMHO you're no worse off than prior to the destroy attempt provided it's clearly reported.

**Comment by ChiefAlexander** [comment]
How do you feel about implementing a flag that by default would keep the current functionality but if enabled would allow you to delete these instances in this error state? Do you feel like that would be a good enough solution while we figure out the implications of cleaning up automatically?
More then happy to try and tackle this if you don't have time @jtopjian

**Comment by jtopjian** [comment]
@ChiefAlexander I like that idea a lot.

We already have options of `force_delete` and `stop_before_destroy`, so how about `destroy_on_error`?

The difference between "delete" and "destroy" is minor but important. `force_delete` points to an actual OpenStack API call called "Force Delete" while "destroy" relates to the Terraform action.

**Comment by jtopjian** [comment]
Hi all,

Per conversations in #427 and #428, it turned out that adding an option would not work. Because of that, I considered this a bug and so instances can now cleanly transition to an error state and then be cleanly deleted.

This was merged in #428 and will be available in the next release.

**Comment by flaper87** [comment]
@jtopjian I believe that was the right thing to do. Thanks for considering it and working on the patch. Great work!

Now carry on, I just wanted to provide some positive feedback :cake:
